store/store.go

```diff
@@ -556,6 +557,19 @@ func (s *Store) refreshDeviceSession(device *auth.DeviceState) error {
 		return fmt.Errorf("internal error: no device and auth context")
 	}
 
+	s.sessionMu.Lock()
+	defer s.sessionMu.Unlock()
+	// check that no other goroutine has already got a new session etc...
+	device1, err := s.dauthCtx.Device()
+	if err != nil {
+		return err
+	}
+	if *device1 != *device {
```

Comment:
How do we know the other device is more recent?

Reply:
I'm not sure I understand the question, Device and UpdateDeviceAuth (and the underlying SetDevice) require/use the global state lock (maybe that was not clear?), so the reading/setting values have a total order, and device1 cannot come before device in that order.

There are actually not many places in the code doing SetDevice and auth.device in state is expected to evolve as:

() -> (brand, model) -> (brand,model,dev-key-id,serial) |1-> (brand,model,dev-key-id,serial, session0) -> (..., session1) -> (..., session2)  etc

This code will not be reached unless we are at least at |1.  There is one Store in snapd and we assume that; before this change  what could happen is that two different thread of execution could use the Store and detect a stale session1, then get from the store session2a and session2b (so the store could possibly think the same device would have two sessions), then one of the SetDevice/UpdateDeviceAuth would be the last and win. But if would do at least some request(s) with the other as well.

For device1 to be different from device there needs to be a SetDevice in between, that can only come from the UpdateDeviceAuth further down in the serialized section by the sessionMu mutex.

Reply:
I think this would make an excellent comment in the tree.